src/include/duckdb/common/cycle_counter.hpp

```diff
@@ -10,32 +10,72 @@
 
 #include "duckdb/common/helper.hpp"
 #include "duckdb/common/chrono.hpp"
+#include "duckdb/common/random_engine.hpp"
 
 namespace duckdb {
 
-//! The cycle counter can be used to measure elapsed cycles
+//! The cycle counter can be used to measure elapsed cycles for a function, expression and ...
+//! Optimized by sampling mechanism. Once per 100 times.
+//! //Todo Can be optimized further by calling RDTSC once per sample
 class CycleCounter {
 public:
-	//! Starts the timer
-	void Start() {
-		finished = false;
-		start = Tick();
+	CycleCounter() : random(0) {
 	}
-	//! Finishes timing
-	void End() {
-		end = Tick();
-		finished = true;
+	// Next_sample determines if a sample needs to be taken, if so start the profiler
+	void TakeSample() {
```

Comment:
Perhaps call these functions BeginSample and EndSample, to make it more obvious that they need to be called together and in that order.